taichi/ir/type.h

```diff
@@ -201,28 +202,38 @@ class TensorType : public Type {
   Type *element_{nullptr};
 };
 
+struct StructMember {
+  const Type *type;
+  std::string name;
+  size_t offset{0};
+  bool operator==(const StructMember &other) const {
+    return type == other.type && name == other.name && offset == other.offset;
+  }
+};
+
 class StructType : public Type {
  public:
-  explicit StructType(std::vector<const Type *> elements)
-      : elements_(std::move(elements)) {
+  explicit StructType(const std::vector<StructMember> &elements,
+                      const std::string &layout = "none")
+      : elements_(elements), layout_(layout) {
   }
 
   std::string to_string() const override;
 
   Type *get_element_type(const std::vector<int> &indices) const;
-  const std::vector<const Type *> &elements() const {
+  const std::vector<StructMember> &elements() const {
     return elements_;
   }
 
   int get_num_elements() const {
     int num = 0;
-    for (const auto &element : elements_) {
-      if (auto struct_type = element->cast<StructType>()) {
+    for (const auto &[type, _, __] : elements_) {
```

Comment:
nit: use `const auto &element` and then `element.type` afterwards since using both `_` and `__` in a structured binding looks quite ugly